wp/lib/bap_wp/src/constraint.ml

```diff
@@ -209,18 +209,43 @@ let rec eval_aux ?stats:(stats = init_stats) (constr : t) (olds : z3_expr list)
       (eval_aux ~stats:stats c2 olds news ctx)
   | Clause (hyps, concs) ->
     let eval_conjunction conj =
-      List.map conj ~f:(fun c -> eval_aux ~stats:stats c olds news ctx)
-      |> Bool.mk_and ctx
+      if List.length conj = 1 then 
+        eval_aux ~stats:stats (List.hd_exn conj) olds news ctx (* This is tail recursive. Avoids And node. *)
+      else
+        List.map conj ~f:(fun c -> eval_aux ~stats:stats c olds news ctx)
+        |> Bool.mk_and ctx
     in
-    let concs_expr = eval_conjunction concs in
     if List.is_empty hyps then
-      concs_expr
+      eval_conjunction concs(* possibly tail recursive? *)
     else
+      let concs_expr = eval_conjunction concs in
       let hyps_expr = eval_conjunction hyps in
       Bool.mk_implies ctx hyps_expr concs_expr
   | Subst (c, o, n) ->
     let n' = List.map n ~f:(fun x -> Expr.substitute x olds news) in
-    eval_aux ~stats:stats c (olds @ o) (news @ n') ctx
+    if (List.length n') = 1 (* It appears we generate overwhelmingly size 1 lists *)
+      then 
+      begin
+        let o = List.hd_exn o in
+        let n' = List.hd_exn n' in
+        (* A Map was tried. It was found that shuffling in and out to lists for Expr.substitute made it not worth it  *)
+        (* The idea here is that popping items most recently used to the front of the list makes it fast to find the most commonly used variables *)
+        (* It is assumed as an invariant that the olds list does not contain duplicate keys *)
+        let findn e l =
+            let rec findn' n e l = match l with | x :: xs ->  if Expr.equal x e then Some n else findn' (n + 1) e xs | [] -> None in
```

Comment:
What does `n`, `e`, and `l` stand for here?

Reply:
e is the expression one is searching for in list l. n is a carried accumulator parameter. It is needed to give the index at which one finds e. Every time e isn't found it is incremented.

Reply:
Do you think we can add this as a comment because I won't remember this after a month?